bfvmm/src/vmm/src/vmm_intel_x64.cpp

```diff
@@ -19,12 +19,490 @@
+
+    if (m_intrinsics == 0 || m_memory_manager == 0)
+        return vmm_error::failure;
+
+    // The following process is documented in the Intel Software Developers
+    // Manual, Section 31.5 Setup VMM & Teardown.
+
+    ret = verify_cpuid_vmx_supported();
+    if (ret != vmm_error::success)
+        return ret;
+
+    ret = verify_vmx_capabilities_msr();
+    if (ret != vmm_error::success)
+        return ret;
+
+    ret = create_vmxon_region();
+    if (ret != vmm_error::success)
+        return ret;
+
+    ret = verify_ia32_vmx_cr0_fixed0_msr();
+    if (ret != vmm_error::success)
+        return ret;
+
+    ret = verify_ia32_vmx_cr0_fixed1_msr();
+    if (ret != vmm_error::success)
+        return ret;
+
+    ret = enable_vmx_operation();
+    if (ret != vmm_error::success)
+        return ret;
+
+    ret = verify_ia32_vmx_cr4_fixed0_msr();
+    if (ret != vmm_error::success)
+        return ret;
+
+    ret = verify_ia32_vmx_cr4_fixed1_msr();
+    if (ret != vmm_error::success)
+        return ret;
+
+    ret = verify_ia32_feature_control_msr();
+    if (ret != vmm_error::success)
+        return ret;
+
+    // The following are additional checks that are documented in the VMXON
+    // instructions itself in the Intel Software Developers Manual,
+    // Section 30.3
+
+    ret = verify_v8086_disabled();
+    if (ret != vmm_error::success)
+        return ret;
+
+    // Finally, execute the VMX on instruction to enter VMX operation.
+    // This places the host OS (code that is currently running) in VMX-root.
+    // The next step will be to launch the VMCS for the host OS, leaving only
+    // the VMM in VMX-root, and placing the host OS into VMX-nonroot.
+
+    ret = execute_vmxon();
+    if (ret != vmm_error::success)
+        return ret;
+
+    return vmm_error::success;
+}
+
+vmm_error::type
+vmm_intel_x64::stop()
+{
+    vmm_error::type ret;
+
+    if (m_intrinsics == 0 || m_memory_manager == 0)
+        return vmm_error::failure;
+
+    ret = execute_vmxoff();
+    if (ret != vmm_error::success)
+        return ret;
+
+    ret = disable_vmx_operation();
+    if (ret != vmm_error::success)
+        return ret;
+
+    ret = release_vmxon_region();
+    if (ret != vmm_error::success)
+        return ret;
+
+    return vmm_error::success;
+}
+
+vmm_error::type
+vmm_intel_x64::create_vmxon_region()
+{
+    if (m_memory_manager->alloc_page(&m_vmxon_page) != memory_manager_error::success)
+        return vmm_error::out_of_memory;
+
+    if (m_vmxon_page.size() < vmxon_vmcs_region_size())
+    {
+        std::cout << "create_vmxon_region failed: "
+                  << "the allocated page is not large enough:" << std::endl
+                  << "    - page size: " << m_vmxon_page.size() << " "
+                  << "    - vmxon/vmcs region size: " << vmxon_vmcs_region_size()
+                  << std::endl;
+        return vmm_error::not_supported;
+    }
+
+    if (((uintptr_t)m_vmxon_page.phys_addr() & 0x0000000000000FFF) != 0)
```

Comment:
This could probably be a little less C like:
m_vmxon_page.phys_addr().isPageAligned() or similiar.